sys/net/gnrc/link_layer/gomach/include/gomach_internal.h

```diff
@@ -0,0 +1,945 @@
+ */
+static inline bool gnrc_gomach_get_got_preamble_ack(gnrc_netif_t *netif)
+{
+    return (netif->mac.prot.gomach.gomach_info & GNRC_GOMACH_INTERNAL_INFO_GOT_PREAMBLEACK);
+}
+
+/**
+ * @brief Set the @ref GNRC_GOMACH_INTERNAL_INFO_ON_PUBCHAN_1 flag of the device.
+ *
+ * @param[in,out] netif      the network interface.
+ * @param[in] on_pubchan_1   value for GoMacH's
+ *                           @ref GNRC_GOMACH_INTERNAL_INFO_ON_PUBCHAN_1 flag.
+ *
+ */
+static inline void gnrc_gomach_set_on_pubchan_1(gnrc_netif_t *netif, bool on_pubchan_1)
+{
+    if (on_pubchan_1) {
+        netif->mac.prot.gomach.gomach_info |= GNRC_GOMACH_INTERNAL_INFO_ON_PUBCHAN_1;
+    }
+    else {
+        netif->mac.prot.gomach.gomach_info &= ~GNRC_GOMACH_INTERNAL_INFO_ON_PUBCHAN_1;
+    }
+}
+
+/**
+ * @brief Get the @ref GNRC_GOMACH_INTERNAL_INFO_ON_PUBCHAN_1 flag of the device.
+ *
+ * @param[in] netif    the network interface.
+ *
+ * @return             true if node is on public channel 1.
+ * @return             false if node is not on public channel 1.
+ */
+static inline bool gnrc_gomach_get_on_pubchan_1(gnrc_netif_t *netif)
+{
+    return (netif->mac.prot.gomach.gomach_info & GNRC_GOMACH_INTERNAL_INFO_ON_PUBCHAN_1);
+}
+
+/**
+ * @brief Set the @ref GNRC_GOMACH_INTERNAL_INFO_MAX_PREAM_INTERV flag of the device.
+ *
+ * @param[in,out] netif  the network interface.
+ * @param[in] max        value for GoMacH's @ref
+ *                       GNRC_GOMACH_INTERNAL_INFO_MAX_PREAM_INTERV flag.
+ *
+ */
+static inline void gnrc_gomach_set_max_pream_interv(gnrc_netif_t *netif, bool max)
+{
+    if (max) {
+        netif->mac.prot.gomach.gomach_info |= GNRC_GOMACH_INTERNAL_INFO_MAX_PREAM_INTERV;
+    }
+    else {
+        netif->mac.prot.gomach.gomach_info &= ~GNRC_GOMACH_INTERNAL_INFO_MAX_PREAM_INTERV;
+    }
+}
+
+/**
+ * @brief Get the @ref GNRC_GOMACH_INTERNAL_INFO_MAX_PREAM_INTERV flag of the device.
+ *
+ * @param[in] netif    the network interface.
+ *
+ * @return             true if node has reached maximum preamble interval.
+ * @return             false if node hasn't reached maximum preamble interval yet.
+ */
+static inline bool gnrc_gomach_get_max_pream_interv(gnrc_netif_t *netif)
+{
+    return (netif->mac.prot.gomach.gomach_info & GNRC_GOMACH_INTERNAL_INFO_MAX_PREAM_INTERV);
+}
+
+/**
+ * @brief Get device's current phase.
+ *
+ * @param[in] netif    the network interface.
+ *
+ * @return             device's current phase.
+ */
+uint64_t gnrc_gomach_phase_now(gnrc_netif_t *netif);
+
+/**
+ * @brief Shortcut to set the state of netdev
+ *
+ * @param[in]   netif       ptr to the network interface
+ * @param[in]   devstate    new state for netdev
+ */
+static inline void gnrc_gomach_set_netdev_state(gnrc_netif_t *netif, netopt_state_t devstate)
+{
+    assert(netif != NULL);
+
+    netif->dev->driver->set(netif->dev,
+                            NETOPT_STATE,
+                            &devstate,
+                            sizeof(devstate));
```

Comment:
`examples/gnrc_networking_mac` and `tests/gomach` block on this line on both `samr21-xpro` and `iotlab-m3` coming from `gomach_listen_init()` (because it got stuck [here](https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx_getset.c#L409)).

Reply:
I get blocked too during my own test,, does it means that we have to enable auto-ACK before using `set_state`??

Reply:
I don't know, my knowledge on network devices isn't extensive enough... Did it ever work? If yes, try to find out what changed state-wise since then.

Reply:
Em,, I see, maybe the radio need some cooling time, I used to get in this sutiation before, let me test it a little bit. and will update soon.

Reply:
BTW, is there recently some radio-driver updates merged into RIOT?

Reply:
This PR used to work just few days ago,, before I rebased on the very recent master branch.

Reply:
Could b283b7784c6391 have broken it? Try with

```sh
git revert b283b7784c6391
```

(but **don't push it**!!1!)

Reply:
Just to be secure, I should do:
1) `git revert b283b7784c6391;`
2) test with nodes;
3) and then? tell you the results?

Reply:
Em, I think the revisions on radio driver (maybe b283b7784c6391) must affect this! After I disable radio sleep, GoMacH will not get bloced,,, it used not to be like this,, something at the lower level has changed.

Reply:
3. See if it works (and report the results, if you like ;-) I'll test myself as soon as I find the time)

Reply:
OK, I'm testing it, but I couldn't revert..:
```
~/Work/RIOT/RIOT $  git revert b283b7784c6391
error: could not revert b283b77... make: fix various compile errors with Wextra
hint: after resolving the conflicts, mark the corrected paths
hint: with 'git add <paths>' or 'git rm <paths>'
hint: and commit the result with 'git commit'

```

Any tips?

Reply:
No more than the hints by git you already pasted ;-).

Reply:
oh, maybe also `git mergetool`

Reply:
BTW, I found that, by using the latest riot/matser branch, I can not even compile (nothing to do with MACs), e.g., `example/degault` applications, for samr21-xpro, (but it works for iotlab-m3). Can you compile on samr21-xpro on the latest master branch (it seems like it works for you). 
But, I have also try in IoTLAB, to make `example/degault`, but failed:
```
zhuo@saclay:~/RIOT/examples/default$ BOARD=samr21-xpro make all
```
```
"make" -C /senslab/users/zhuo/RIOT/sys/shell/commands
"make" -C /senslab/users/zhuo/RIOT/sys/tsrb
"make" -C /senslab/users/zhuo/RIOT/sys/uart_stdio
"make" -C /senslab/users/zhuo/RIOT/sys/xtimer
cortexm.ld:25: nonconstant expression for origin
collect2: error: ld returned 1 exit status
/senslab/users/zhuo/RIOT/examples/default/../../Makefile.include:320: recipe for target 'link' failed
make: *** [link] Error 1
```
So, do you know what happen to `RIOT/Makefile.include:320` ? :-/

Reply:
Didn't you already report that in https://github.com/RIOT-OS/RIOT/pull/5618#issuecomment-347633739? I don't have this issue.... Which toolchain are you using?

Reply:
yes, it is that report. I used gcc-arm-none-eabi (maybe 4.9.3)... But, I have the same issue in remote making in IoTLAB...

BTW, could I bother you for testing this PR and when you have time.  It seems that so many changes have been made to the master, and I can hardly know why.. Thanks in advance!

Reply:
I tried now to revert every major change to the device driver since October, but it did not solve the issue :-/...

Reply:
OK, thanks!, let me test also..

Reply:
With the state of the driver reverted to the state before https://github.com/RIOT-OS/RIOT/pull/7309 it works now. I'll investigate, but I think you are doing something you are not supposed to here ...

Reply:
Great!~
> you are doing something you are not supposed to here ...

What do you refer to?

Reply:
(directly after #7309 it was working as well)

Reply:
> What do you refer to?

I don't know yet. That's why I said "I'll investigate".

Reply:
OK, what I do with GoMacH is to:
1)  turn-on turn-off radio;
2) turn the radio to a specific channel;
3) enable or disable auto-ACK;
4) set ACK-require.
That is all~

Reply:
Anyway, thanks a lot!~ Martine

Reply:
I was able to pin-point the regression to 961c49e2cc138e1ff2089eec173d692122884ec2. @smlng any idea which might cause this?

Reply:
So it isn't your fault after all ;-)

Reply:
Because this simple fix

```diff
diff --git a/drivers/at86rf2xx/at86rf2xx_getset.c b/drivers/at86rf2xx/at86rf2xx_getset.c
index cb3e14f..9824586 100644
--- a/drivers/at86rf2xx/at86rf2xx_getset.c
+++ b/drivers/at86rf2xx/at86rf2xx_getset.c
@@ -435,7 +435,10 @@ uint8_t at86rf2xx_set_state(at86rf2xx_t *dev, uint8_t state)
     }
     else if (state != old_state) {
         /* we need to go via PLL_ON if we are moving between RX_AACK_ON <-> TX_ARET_ON */
-        if ((old_state | state) == (AT86RF2XX_STATE_RX_AACK_ON | AT86RF2XX_STATE_TX_ARET_ON)) {
+        if ((old_state == AT86RF2XX_STATE_RX_AACK_ON &&
+                         state == AT86RF2XX_STATE_TX_ARET_ON) ||
+                     (old_state == AT86RF2XX_STATE_TX_ARET_ON &&
+                          state == AT86RF2XX_STATE_RX_AACK_ON)) {
             _set_state(dev, AT86RF2XX_STATE_PLL_ON, AT86RF2XX_STATE_PLL_ON);
         }
         /* check if we need to wake up from sleep mode */
```

makes it disappear.

Reply:
Will provide a fix to master, I think we don't need to rebase though.

Reply:
AH!~ Thanks a lot!!! Martine!~

Reply:
> Will provide a fix to master, I think we don't need to rebase though.

? I should rebase to the master after you fix it, right (your fix PR merged into master)?

Reply:
You can do so if you want to. It is already ACK'd so this would be a thing of minutes.

Reply:
It's merged, so if you want to: go ahead and rebase.

Reply:
Huge thanks!~ Martine~